src/org/labkey/test/components/react/QueryChartDialog.java

```diff
@@ -0,0 +1,330 @@
+                "the Save chart button did not become enabled", 2000);
+        dismiss("Save Chart");
+        return _queryGrid.getChartPanel();
+    }
+
+    /*
+        appears when dialog is in 'edit' mode
+     */
+    public void clickDeleteChart(boolean confirmDelete)
+    {
+        // clicking 'delete chart' shows a prompt + cancel/delete buttons
+        var dismissBtn = Locators.dismissButton("Delete Chart").waitForElement(getComponentElement(), 1500);
+        getWrapper().shortWait().until(ExpectedConditions.elementToBeClickable(dismissBtn));
+        dismissBtn.click();
+        getWrapper().shortWait().until(ExpectedConditions.stalenessOf(dismissBtn));
+
+        // warn div contains cancel and delete buttons
+        var warnDiv = Locator.tagWithClass("div", "form-buttons__right")
+                .containing("Are you sure you want to permanently delete this chart?")
+                .waitForElement(this, 1000);
+        if (confirmDelete)
+        {
+            Locators.dismissButton("Delete").waitForElement(warnDiv, 500).click();
+            waitForClose();
+        }
+        else
+        {
+            var cancelBtn = Locators.dismissButton("Cancel").waitForElement(warnDiv, 500);
+            cancelBtn.click();
+            getWrapper().shortWait().until(ExpectedConditions.stalenessOf(cancelBtn));
+        }
+    }
+
+    @Override
+    protected ElementCache newElementCache()
+    {
+        return new ElementCache();
+    }
+
+    @Override
+    protected ElementCache elementCache()
+    {
+        return (ElementCache) super.elementCache();
+    }
+
+    protected class ElementCache extends ModalDialog.ElementCache
+    {
+        final Input nameInput = Input(Locator.input("name"), getDriver()).findWhenNeeded(this);
+        final Checkbox sharedCheckbox = Checkbox.Checkbox(Locator.input("shared")).findWhenNeeded(this);
+
+        Locator.XPathLocator chartBuilderType = Locator.tagWithClass("div", "chart-builder-type");
+        public CHART_TYPE getSelectedChartType()
+        {
+            var selectedEl = chartBuilderType.withAttributeContaining("class", "selected")
+                    .waitForElement(this, 1500);
+            String dataName = selectedEl.getAttribute("data-name");
+            return CHART_TYPE.fromChartType(dataName);
+        }
+
+        // note: selecting a chart type will cause most select fields to go stale
+        public void selectChartType(CHART_TYPE chartType)
+        {
+            if (getSelectedChartType().equals(chartType))
+                return;
+
+            var el = chartBuilderType.withAttribute("data-name", chartType.getChartType())
+                    .waitForElement(this, 1500);
+            getWrapper().shortWait().until(ExpectedConditions.elementToBeClickable(el));
+            el.click();
+            WebDriverWrapper.waitFor(()-> getSelectedChartType().equals(chartType),
+                    "The requested chart type did not become selected", 2000);
+        }
```

Comment:
These aren't really elements, these are methods. They shouldn't be in the element cache.